Handling mount during recovery for an already attached volume

pshahzeb opened the pull request:
## Description
When the plugin is disabled with active containers with mounted volumes, plugin rootfs is gone but volume is still attached and namespace mounted to the container. In such a case the device path is present on the host.

When the plugin comes up, this case needs to be handled and volume mounted in plugin rootfs

This PR makes mount during plugin recovery resilient. An already attached but not mounted in rootfs volume is mounted during plugin recovery.

Do not wait for dev attach if the device is already attached. This ensures plugin doesn't endup in timeout.

The recovery mount takes advantage of facts:
1. if a volume is attached to a VM, the esx service returns the correct device ids to the plugin
2. When such a volume is attached but not mounted in plugin rootfs, we mount it and use that as mount point for next .

## Testing
Done manually with docker 17.06 while containers are running with --restart flag.
See the detailed steps below for a combination of upgrade and data loss testing.

```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin ls
ID                  NAME                DESCRIPTION         ENABLED


root@sc-rdops-vm02-dhcp-52-237:~# docker plugin install --grant-all-permissions --alias vsphere vmware/docker-volume-vsphere:latest VDVS_LOG_LEVEL=debug
latest: Pulling from vmware/docker-volume-vsphere
ad10a6bbce92: Download complete
Digest: sha256:feaa90e4234880fbf1669015e88cd485b2077270219ff84a730b35207bb622d4
Status: Downloaded newer image for vmware/docker-volume-vsphere:latest
Installed plugin vmware/docker-volume-vsphere:latest
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create -d vsphere --name vol1
vol1

root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere:latest      vol1@sharedvmfs-0


root@sc-rdops-vm02-dhcp-52-237:~# docker run -d --name con_1 -v vol1@sharedvmfs-0:/data alpine ping 8.8.8.8
50150ed538e6e14e8033acdc2325715ee3939e9ba342bb39cae3a43b3c7e0f6e



root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_1 /bin/sh
/ # echo "Stage 1 " > /data/text
/ # cat /data/text
Stage 1
/ # df -h /data
Filesystem                Size      Used Available Use% Mounted on
/dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:0:0
                         92.8M      1.5M     84.3M   2% /data
/ # exit


root@sc-rdops-vm02-dhcp-52-237:~# docker plugin disable -f vsphere
vsphere
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin ls
ID                  NAME                DESCRIPTION                           ENABLED
097eea81301d        vsphere:latest      VMWare vSphere Docker Volume plugin   false
root@sc-rdops-vm02-dhcp-52-237:~# cat /proc/mounts | grep vol1
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#

root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_1 /bin/sh
/ # cat /data/text
Stage 1
/ # echo "Stage 1, Stage 2 " > /data/text
/ # cat /data/text
Stage 1, Stage 2
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~#


root@sc-rdops-vm02-dhcp-52-237:~# docker plugin upgrade vsphere:latest pshahzeb/docker-volume-vsphere:0.16-dev
Upgrading plugin vsphere:latest from vmware/docker-volume-vsphere:latest to pshahzeb/docker-volume-vsphere:0.16-dev
Plugin images do not match, are you sure? [y/N] y
Plugin "pshahzeb/docker-volume-vsphere:0.16-dev" is requesting the following privileges:
 - mount: [/dev]
 - mount: [/var/run]
 - mount: [/etc]
 - mount: [/var/log]
 - allow-all-devices: [true]
 - capabilities: [CAP_SYS_ADMIN]
Do you grant the above permissions? [y/N] y
0.16-dev: Pulling from pshahzeb/docker-volume-vsphere
b4282b72a544: Download complete
Digest: sha256:aa2ebde6d348786893a186415bf68722b0c4234a8775f0939331ba25ac2c74b6
Status: Downloaded newer image for pshahzeb/docker-volume-vsphere:0.16-dev
Upgraded plugin vsphere:latest to pshahzeb/docker-volume-vsphere:0.16-dev
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin ls
ID                  NAME                DESCRIPTION                           ENABLED
097eea81301d        vsphere:latest      VMWare vSphere Docker Volume plugin   false
root@sc-rdops-vm02-dhcp-52-237:~# cat /proc/mounts | grep vol1


root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_1 /bin/sh
/ # cat /data/text
Stage 1, Stage 2
/ # echo "Stage 1, Stage 2, Stage 3 " > /data/text
/ # cat /data/text
Stage 1, Stage 2, Stage 3
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~#


vsphere
root@sc-rdops-vm02-dhcp-52-237:~# cat /proc/mounts | grep vol1
/dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:0:0 /var/lib/docker/plugins/097eea81301d776a15c2025edf3601f8078b5b292625446cb7b178719f37ae3d/rootfs/mnt/vmdk/vol1@sharedvmfs-0 ext4 rw,relatime,data=ordered 0 0
/dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:0:0 /var/lib/docker/plugins/097eea81301d776a15c2025edf3601f8078b5b292625446cb7b178719f37ae3d/propagated-mount/vol1@sharedvmfs-0 ext4 rw,relatime,data=ordered 0 0


root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_1 /bin/sh
/ # cat /data/text
Stage 1, Stage 2, Stage 3
/ # echo "Stage 1, Stage 2, Stage 3, Stage 4 " > /data/text
/ # cat /data/text
Stage 1, Stage 2, Stage 3, Stage 4
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~#

root@sc-rdops-vm02-dhcp-52-237:~# docker run -d --name con_2 -v vol1@sharedvmfs-0:/data alpine ping 8.8.8.8
68e64b4d168855634d35b6c4407a176da73055cc4721ab221855025f359cf32c
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_2 /bin/sh
/ # cat /data/text
Stage 1, Stage 2, Stage 3, Stage 4
/ # echo "Stage 1, Stage 2, Stage 3, Stage 4, Stage 5 " > /data/text
/ # cat /data/text
Stage 1, Stage 2, Stage 3, Stage 4, Stage 5
/ # exit


root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_1 /bin/sh
/ # cat /data/text
Stage 1, Stage 2, Stage 3, Stage 4, Stage 5
/ # echo "Stage 1, Stage 2, Stage 3, Stage 4, Stage 5, Stage 6 " > /data/text
/ # cat /data/text
Stage 1, Stage 2, Stage 3, Stage 4, Stage 5, Stage 6
/ # exit
```

Fixes #1779

pshahzeb commented:
> I only see the " do not wait for existing file" portion (which looks good) . Was there supposed to be anything else ?

With just this fix, I think all cases are getting handled. The mount during recovery was already there, but appropriately using the device file and not running into timeout handles solves all issues. I feel all other stuff (getting device info for each container from docker, handling  mounts calls) aren't needed anymore.

msterin commented:
Cool. Don't we have to remove useless code then ? (in another PR) ?

msterin commented:
@pshahzeb - so why it did not work before ? What sis the time out broke, other than slow response ?

pshahzeb commented:
>  so why it did not work before ? What sis the time out broke, other than slow response ?
Earlier the slow response got the plugin not to start listening to the socket.

While removing the dependency on /proc/mounts,  I removed the mounts during recovery as well, which caused the later containers coming up to not have a correct mount until the already running container exited. (should have kept earlier commit, I rebased the commits)

Since we only needed the device file of device already attached, and we already had a recovery mount, keeping the recovery in mounting in place and just fixing the timeout fixes all the problem.

pshahzeb commented:
@govint thanks for suggesting this test.
As of now I have tested these steps manually. All good. See detailed steps below
```
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin install --grant-all-permissions --alias vsphere vmware/docker-volume-vsphere:latest VDVS_LOG_LEVEL=debug
latest: Pulling from vmware/docker-volume-vsphere
ad10a6bbce92: Download complete
Digest: sha256:feaa90e4234880fbf1669015e88cd485b2077270219ff84a730b35207bb622d4
Status: Downloaded newer image for vmware/docker-volume-vsphere:latest
Installed plugin vmware/docker-volume-vsphere:latest
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin ls
ID                  NAME                DESCRIPTION                           ENABLED
b976205b9d47        vsphere:latest      VMWare vSphere Docker Volume plugin   true
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create -d vsphere --name vol1
vol1
root@sc-rdops-vm02-dhcp-52-237:~# docker run -d --name con_1 -v vol1@sharedVmfs-0:/data alpine ping 8.8.8.8
ecb53a934873925e2a89078049084d502c17646d20bf4307a7abcd28878d2358
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_1 /bin/sh
/ # ^C
/ # echo "Stage 1 " >> /data/text
/ # cat /data/text
Stage 1
/ # df -h
Filesystem                Size      Used Available Use% Mounted on
none                      2.9G      2.0G    783.3M  72% /
tmpfs                     1.9G         0      1.9G   0% /dev
tmpfs                     1.9G         0      1.9G   0% /sys/fs/cgroup
/dev/disk/by-path/pci-0000:03:00.0-scsi-0:0:0:0
                         92.8M      1.5M     84.3M   2% /data
/dev/disk/by-uuid/66a21986-55c4-4ede-a2d9-37d30aa41a8e
                          2.9G      2.0G    783.3M  72% /etc/resolv.conf
/dev/disk/by-uuid/66a21986-55c4-4ede-a2d9-37d30aa41a8e
                          2.9G      2.0G    783.3M  72% /etc/hostname
/dev/disk/by-uuid/66a21986-55c4-4ede-a2d9-37d30aa41a8e
                          2.9G      2.0G    783.3M  72% /etc/hosts
shm                      64.0M         0     64.0M   0% /dev/shm
tmpfs                     1.9G         0      1.9G   0% /proc/kcore
tmpfs                     1.9G         0      1.9G   0% /proc/timer_list
tmpfs                     1.9G         0      1.9G   0% /proc/timer_stats
tmpfs                     1.9G         0      1.9G   0% /proc/sched_debug
tmpfs                     1.9G         0      1.9G   0% /sys/firmware
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin disable -f vsphere
vsphere
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin disable -f vsphere^C
root@sc-rdops-vm02-dhcp-52-237:~# sudo apt-get updaet^C
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin upgrade vsphere:latest pshahzeb/docker-volume-vsphere:0.16-dev
Upgrading plugin vsphere:latest from vmware/docker-volume-vsphere:latest to pshahzeb/docker-volume-vsphere:0.16-dev
Plugin images do not match, are you sure? [y/N] y
Plugin "pshahzeb/docker-volume-vsphere:0.16-dev" is requesting the following privileges:
 - mount: [/dev]
 - mount: [/var/run]
 - mount: [/etc]
 - mount: [/var/log]
 - allow-all-devices: [true]
 - capabilities: [CAP_SYS_ADMIN]
Do you grant the above permissions? [y/N] y
0.16-dev: Pulling from pshahzeb/docker-volume-vsphere
b4282b72a544: Download complete
Digest: sha256:aa2ebde6d348786893a186415bf68722b0c4234a8775f0939331ba25ac2c74b6
Status: Downloaded newer image for pshahzeb/docker-volume-vsphere:0.16-dev
Upgraded plugin vsphere:latest to pshahzeb/docker-volume-vsphere:0.16-dev
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin enable vsphere
vsphere
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_1 /bin/sh
/ # cat /data/text
Stage 1
/ # exit
(failed reverse-i-search)`ryb': docke^Cexec -it con_1 /bin/sh
root@sc-rdops-vm02-dhcp-52-237:~# docker run -d --name con_2 -v vol1@sharedVmfs-0:/data alpine ping 8.8.8.8
5151cc5177e10431152633e05bd3233cd60dd8125116084cb3c6df697bc05660
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_2 /bin/sh
/ # cat /data/text
Stage 1
/ # echo "Stage 2 " >> /data/text
/ # cat /data/text
Stage 1
Stage 2
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_1 /bin/sh
/ # cat /data/text
Stage 1
Stage 2
/ # echo "Stage 3" >>  /data/text
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin disable
"docker plugin disable" requires exactly 1 argument(s).
See 'docker plugin disable --help'.

Usage:  docker plugin disable [OPTIONS] PLUGIN

Disable a plugin
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin disable -f vsphere
vsphere
root@sc-rdops-vm02-dhcp-52-237:~# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
5151cc5177e1        alpine              "ping 8.8.8.8"      4 minutes ago       Up 4 minutes                            con_2
ecb53a934873        alpine              "ping 8.8.8.8"      5 minutes ago       Up 5 minutes                            con_1
root@sc-rdops-vm02-dhcp-52-237:~# docker stop `docker ps -q`

5151cc5177e1
ecb53a934873
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~#
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin enable vsphere
vsphere
root@sc-rdops-vm02-dhcp-52-237:~# docker run -d --name con_3 -v vol1@sharedVmfs-0:/data alpine ping 8.8.8.8
dca8024146dea4604f9758a32bcafc0897afb5245ee1bb5b9965338be00a676c
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_3 /bin/sh
/ # cat /data/text
Stage 1
Stage 2
Stage 3
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_3 /bin/sh
/ # echo "Stage 4 " >> /data/text
/ # cat /data/text
Stage 1
Stage 2
Stage 3
Stage 4
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin disable -f vsphere
vsphere
root@sc-rdops-vm02-dhcp-52-237:~# service docker restart
docker stop/waiting
docker start/running, process 8390
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin enable vsphere
vsphere
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_4 /bin/sh
Error: No such container: con_4
root@sc-rdops-vm02-dhcp-52-237:~# ^C
root@sc-rdops-vm02-dhcp-52-237:~# docker run -d --name con_4 -v vol1@sharedVmfs-0:/data alpine ping 8.8.8.8
0ff984bc4255fa4aade25013fed65c48960be9ca7e2dc87bf46910d4a739345e
root@sc-rdops-vm02-dhcp-52-237:~# docker exec -it con_4 /bin/sh
/ # cat /data/text
Stage 1
Stage 2
Stage 3
Stage 4
/ # exit
root@sc-rdops-vm02-dhcp-52-237:~#
```
